### PR TITLE
openstack_networking_router: enable/disable snat

### PR DESCRIPTION
**Pull request by Crapworks**
Hi!

This is for #134

This PR adds the `enable_snat` property to the networking_router_v2 resource to enable/disable snat for neutron routers.

I've testet this against my openstack provider and it works fine! But again, I don't really know GO, so if this can be done more elegant, I'm open for suggestions :)

I am also not sure if it is the right thing to raise an error if `external_gateway` is not set, or if I should just ignore the parameter in that case.

Cheers,
Christian

### CODE REVIEW

File: openstack/resource_openstack_networking_router_v2.go [hunk 1]
```diff
@@ -186,6 +201,14 @@ func resourceNetworkingRouterV2Update(d *schema.ResourceData, meta interface{})
...
 		}
 	}
+	if d.HasChange("enable_snat") {
+		enableSNAT := d.Get("enable_snat").(bool)
+		if updateOpts.GatewayInfo != nil {
+			updateOpts.GatewayInfo.EnableSNAT = &enableSNAT
+		} else {
+			return fmt.Errorf("Error setting enable_snat: option requires external_gateway to be set")
+		}
+	}
```
Comment (jtopjian) [inline openstack/resource_openstack_networking_router_v2.go]: There's an issue with the logic here.

Let's say a user starts with this configuration:

```hcl
resource "openstack_networking_router_v2" "router_1" {
  name = "router"                                     
  admin_state_up = "true"                             
  distributed = "false"                               
  external_gateway = "asdf"                             
}                                                     
```
Then

```shell
$ terraform apply
```

And then they decide to disable SNAT:

```hcl
resource "openstack_networking_router_v2" "router_1" {
  name = "router"                                     
  admin_state_up = "true"                             
  distributed = "false"                               
  external_gateway = "asdf"
  enable_snat = false
}                                                     
```

`external_gateway` isn't being updated, so the code within `d.HasChange("external_gateway")` is never triggered. When the user runs another `terraform apply`, they'll get an error.

What about something like this?

```go
var gatewayInfoChanged bool
gatewayInfo := routers.GatewayInfo{}

if d.HasChange("external_gateway") {
	externalGateway := d.Get("external_gateway").(string)
	if externalGateway != "" {
                gatewayInfoChanged = true
                gatewayInfo.NetworkID = externalGateway
	}
}
if d.HasChange("enable_snat") {
        externalGateway := d.Get("external_gateway").(string)
        if externalGateway == "" {
		return fmt.Errorf("Error setting enable_snat: option requires external_gateway to be set")
        }
	enableSNAT := d.Get("enable_snat").(bool)
        gatewayInfoChanged = true
        gatewayInfo.EnableSNAT = &enableSNAT
        gatewayInfo.NetworkID = externalGateway
}

if gatewayInfoChanged {
        updateOpts.GatewayInfo = &gatewayInfo
}
```

Note how `gatewayInfo.NetworkID` is being set in the `enable_snat` block of code. The two must be sent together.

**Comment by Crapworks** [comment]
@jtopjian Oh, you're right. I wasn't testing this scenario. I am just wondering, if I do something like this:

```go
    gatewayInfo := routers.GatewayInfo{}
    externalGateway := d.Get("external_gateway").(string)
    if externalGateway != "" { 
        gatewayInfo.NetworkID = externalGateway
    } 
    if d.HasChange("external_gateway") { 
        updateOpts.GatewayInfo = &gatewayInfo
    } 
    if d.HasChange("enable_snat") { 
        enableSNAT := d.Get("enable_snat").(bool)
        gatewayInfo.EnableSNAT = &enableSNAT
        if gatewayInfo.NetworkID != "" { 
            updateOpts.GatewayInfo = &gatewayInfo
        } else { 
            return fmt.Errorf("Error setting enable_snat: option requires external_gateway to be set")
        } 
    } 
```

It should work as well, even without having an additional variable that checks for changes, right? Or am I missing something?

**Comment by jtopjian** [comment]
Yep, that works, too :)

**Comment by Crapworks** [comment]
@jtopjian I've made the changes

**Comment by jtopjian** [comment]
Looks good! Thank you for working on this.

**Pull request merged**